sim/paladin/apl_values.go

```diff
@@ -0,0 +1,93 @@
+package paladin
+
+import (
+	"time"
+
+	"github.com/wowsims/sod/sim/core"
+	"github.com/wowsims/sod/sim/core/proto"
+)
+
+// The APLValue for the remaining duration of the primary seal aura.
+
+func (paladin *Paladin) NewAPLValue(rot *core.APLRotation, config *proto.APLValue) core.APLValue {
+	switch config.Value.(type) {
+	case *proto.APLValue_CurrentSealRemainingTime:
+		return paladin.newValueCurrentSealRemainingTime(rot, config.GetCurrentSealRemainingTime())
+	default:
+		return nil
+	}
+}
+
+type APLValueCurrentSealRemainingTime struct {
+	core.DefaultAPLValueImpl
+	paladin *Paladin
+}
+
+func (paladin *Paladin) newValueCurrentSealRemainingTime(rot *core.APLRotation, config *proto.APLValueCurrentSealRemainingTime) core.APLValue {
+	return &APLValueCurrentSealRemainingTime{
+		paladin: paladin,
+	}
+}
+
+func (value *APLValueCurrentSealRemainingTime) Type() proto.APLValueType {
+	return proto.APLValueType_ValueTypeDuration
+}
+
+func (value *APLValueCurrentSealRemainingTime) GetDuration(sim *core.Simulation) time.Duration {
+	paladin := value.paladin
+	return max(paladin.CurrentSealExpiration - sim.CurrentTime)
+}
+
+func (value *APLValueCurrentSealRemainingTime) String() string {
+	return "Current Seal Remaining Time()"
+}
+
+// The APLAction for casting the current Seal
+func (paladin *Paladin) NewAPLAction(rot *core.APLRotation, config *proto.APLAction) core.APLActionImpl {
+	switch config.Action.(type) {
+	case *proto.APLAction_CastPaladinPrimarySeal:
+		return paladin.newActionPaladinPrimarySealAction(rot, config.GetCastPaladinPrimarySeal())
+	default:
+		return nil
+	}
+}
+
+type APLActionCastPaladinPrimarySeal struct {
+	paladin    *Paladin
+	lastAction time.Duration
+}
+
+func (impl *APLActionCastPaladinPrimarySeal) GetInnerActions() []*core.APLAction { return nil }
+func (impl *APLActionCastPaladinPrimarySeal) GetAPLValues() []core.APLValue      { return nil }
+func (impl *APLActionCastPaladinPrimarySeal) Finalize(*core.APLRotation)         {}
+func (impl *APLActionCastPaladinPrimarySeal) GetNextAction(*core.Simulation) *core.APLAction {
+	return nil
+}
+
+func (paladin *Paladin) newActionPaladinPrimarySealAction(_ *core.APLRotation, config *proto.APLActionCastPaladinPrimarySeal) core.APLActionImpl {
+	return &APLActionCastPaladinPrimarySeal{
+		paladin: paladin,
+	}
+}
+
+func (action *APLActionCastPaladinPrimarySeal) Execute(sim *core.Simulation) {
+	action.lastAction = sim.CurrentTime
+	paladin := action.paladin
+	// If the player options are incorrectly configured, then no primary seal will be selected.
+	if paladin.PrimarySealSpell == nil {
+		return
+	}
+	paladin.PrimarySealSpell.Cast(sim, paladin.CurrentTarget)
+}
+
+func (action *APLActionCastPaladinPrimarySeal) IsReady(sim *core.Simulation) bool {
+	return sim.CurrentTime > action.lastAction
```

Comment:
You probably want to add `&& paladin.PrimarySealSpell.CanCast(sim)` so this action is not executed when the spell cant be cast